pkg/admissionpolicy/mutate.go

```diff
@@ -112,23 +126,310 @@ func mutateResource(
 			Namespace:           namespace,
 			TypeConverter:       managedfields.NewDeducedTypeConverter(),
 		}
+		original := versionedAttributes.VersionedObject
+
 		newVersionedObject, err := patcher.Patch(context.TODO(), patchRequest, celconfig.RuntimeCELCostBudget)
 		if err != nil {
-			return engineResponse, nil
+			ruleResp := engineapi.RuleError(policy.GetName(), engineapi.Mutation, err.Error(), nil, nil)
+			policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
+			return engineResponse.WithPolicyResponse(policyResp), nil
+		}
+
+		// check if mutation actually changed anything
+		if equality.Semantic.DeepEqual(original, newVersionedObject) {
+			ruleResp := engineapi.RuleSkip(policy.GetName(), engineapi.Mutation, "mutation had no effect", nil)
+			policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
+			return engineResponse.WithPolicyResponse(policyResp), nil
 		}
-		versionedAttributes.Dirty = true
+
 		versionedAttributes.VersionedObject = newVersionedObject
+
 	}
+
 	patchedResource, err := celutils.ConvertObjectToUnstructured(versionedAttributes.VersionedObject)
 	if err != nil {
-		return engineResponse, err
+		ruleResp := engineapi.RuleError(policy.GetName(), engineapi.Mutation, err.Error(), nil, nil)
+		policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
+		return engineResponse.WithPolicyResponse(policyResp), nil
 	}
 
 	ruleResp := engineapi.RulePass(policy.GetName(), engineapi.Mutation, "", nil)
+	if binding != nil {
+		ruleResp = ruleResp.WithMutatingBinding(binding)
+	}
 	policyResp.Add(engineapi.NewExecutionStats(startTime, time.Now()), *ruleResp)
 	engineResponse = engineResponse.
 		WithPatchedResource(*patchedResource).
 		WithPolicyResponse(policyResp)
 
 	return engineResponse, nil
 }
+
+// MutatingPolicyData holds a MAP and its associated MAPBs
+type MutatingPolicyData struct {
+	definition admissionregistrationv1alpha1.MutatingAdmissionPolicy
+	bindings   []admissionregistrationv1alpha1.MutatingAdmissionPolicyBinding
+}
+
+// NewMutatingPolicyData initializes a MAP wrapper with no bindings
+func NewMutatingPolicyData(p admissionregistrationv1alpha1.MutatingAdmissionPolicy) MutatingPolicyData {
+	return MutatingPolicyData{definition: p}
+}
+
+// AddBinding appends a MAPB to the policy data
+func (m *MutatingPolicyData) AddBinding(b admissionregistrationv1alpha1.MutatingAdmissionPolicyBinding) {
+	m.bindings = append(m.bindings, b)
+}
```

Comment:
Can you move it to `pkg/engine/api/policy.go`?
You can find the VAP API there too.
https://github.com/kyverno/kyverno/blob/12e5ca9b4dde3b22c6df1eb498e0188821bc51a6/pkg/engine/api/policy.go#L11-L38